src/messaging/messaging.ts

```diff
@@ -332,6 +339,40 @@ function validateAps(aps: Aps) {
   }
 }
 
+function validateApsSound(sound: string | CriticalSound) {
+  if (typeof sound === 'undefined' || validator.isString(sound)) {
+    return;
+  } else if (!validator.isNonNullObject(sound)) {
+    throw new FirebaseMessagingError(
+      MessagingClientErrorCode.INVALID_PAYLOAD,
+      'apns.payload.aps.sound must be a string or a non-null object');
+  }
+
+  const volume = (sound as CriticalSound).volume;
+  if (typeof volume !== 'undefined') {
+    if (!validator.isNumber(volume)) {
+      throw new FirebaseMessagingError(
+        MessagingClientErrorCode.INVALID_PAYLOAD,
+        'apns.payload.aps.sound.volume must be a number');
+    }
+    if (volume < 0 || volume > 1) {
+      throw new FirebaseMessagingError(
+        MessagingClientErrorCode.INVALID_PAYLOAD,
+        'apns.payload.aps.sound.volume must be in the interval [0, 1]');
+    }
+  }
+  const soundObject = sound as {[key: string]: any};
+  const key = 'critical';
+  const critical = soundObject[key];
+  if (typeof critical !== 'undefined' && critical !== 1) {
+    if (critical === true) {
+      soundObject[key] = 1;
```

Comment:
The name of the function indicates it just validates the input and doesn't have a return value (void), but we modify the input directly. It's hard for people to understand the intention and this style should be avoid. I would rename it to something like `convertToApsSoundPayload` and return the modified object. At least I would put some comments to explain what's going on, but I really think we shouldn't modify input directly.

Reply:
We have a bunch of un-exported helper functions like this in this module. They all directly modify the input. However, we guarantee that these methods receive a copy of the original input, so in-place modification is ok. 

I know it's not ideal, but I want to follow the existing pattern here. We can change them to not mutate the input in a separate PR.